### PR TITLE
Fix silent forward overflow in softplus for reduced precision types

### LINKED ISSUE #187180: torch.nn.functional.softplus returns +Inf for finite float16 input when beta is very small
### 🐛 Describe the bug

`torch.nn.functional.softplus(x, beta)` on a `float16` input returns `+inf` for an ordinary finite input when `beta` is very small, even though the result is finite in `float32`. softplus is `(1/beta) * log1p(exp(beta * x))`. With `beta = 1e-6` and `x = 1.0`, `exp(1e-6) ~= 1.000001` and `log1p(...) ~= 0.693147`, so the value is `0.693147 / 1e-6 ~= 693147.7`. That is finite in fp32 but exceeds the fp16 max of `65504`; the reduced-floating-type kernel widens to fp32 for the math and then does a final `static_cast<scalar_t>(...)` back to fp16 with no range check, overflowing to `+inf`.

The linear stability branch (`beta * x > threshold`, default `20`) never fires here because `beta * x = 1e-6`. This is distinct from the large-`beta` exp-overflow case: here the whole fp32 computation is finite and only the narrowing cast overflows.

```python
import torch
import torch.nn.functional as F

x16 = torch.tensor([[1.0]], dtype=torch.float16)   # finite input
print("fp16:", F.softplus(x16, beta=1e-6).item())
print("fp32:", F.softplus(x16.float(), beta=1e-6).item())
print("fp16 max:", torch.finfo(torch.float16).max)
```

Observed:

```
fp16: inf
fp32: 693147.6875
fp16 max: 65504.0
```

Expected:

Compute in fp32 and clamp to `finfo(float16).max` before the narrowing cast (or raise on overflow), rather than silently returning `+inf` for a finite input.

### Versions

```
PyTorch version: 2.5.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-124-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 535.309.01
cuDNN version: Could not collect
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU(s):                                  20
Model name:                              12th Gen Intel(R) Core(TM) i7-12700
Thread(s) per core:                      2
Core(s) per socket:                      12

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.5.1+cu121
[pip3] triton==3.1.0
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by BhagyeshKothalkar**
Fixes #187180
# Summary & Motivation
Currently, `torch.nn.functional.softplus` causes silent overflow to +inf when used with reduced precision types (float16, bfloat16) on inputs that bypass the linear threshold. This happens because the intermediate transcendental math is computed in float32 (via opmath_t), and results can easily exceed the float16 maximum (65504.0) before being downcast, breaking the mathematical contract of the activation function and poisoning the autograd graph with NaN gradients on the backward pass.

# Solution
This PR introduces a ceiling clamp in the forward kernel to prevent this downcast overflow, while strictly preserving +inf inputs and IEEE 754 NaN propagation. The fix is impllemented in CPU and CUDA kernels. aa test is added in test/test_nn.py

For the CUDA kernels, the clamp is guarded by if constexpr (!std::is_same_v<scalar_t, opmath_t>). This ensures that for full-precision types (float32, float64), the clamp logic and static_cast are entirely optimized away at compile time, maintaining original performance with no code duplication. For CPU kernels, the floats and reduced floats are already dispatched separately.
The initialization of max_val is localized strictly within the if constexpr block to prevent -Wunused-variable warnings. 

# Backward Compatibility
The only behaviour change caused is that on float16 for large numbers, the computation clamps to tghe max of that dtype instead of returning inf. 

# Testing & Verification
I have expanded the test suite to rigorously cover these boundary conditions:

Added tests with large inputs and small betas (e.g., x=60000.0, beta=1e-5) to prove the clamp catches values that bypass the linear threshold (x * beta > 20) but still mathematically exceed dtype_max.
Added specific assertions and comments for bfloat16. Since bfloat16 shares the float32 exponent range, the math path will hit +inf before producing a finite number large enough to trigger the bfloat16 clamp. The test now explicitly verifies that the clamp acts as an intentional, safe no-op for bfloat16 using strict inequalities.

Code Touched
aten/src/ATen/native/cpu/Activation.cpp
aten/src/ATen/native/cuda/ActivationSoftplusKernel.cu
test/test_nn.py

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by BhagyeshKothalkar** [comment]
Closing this draft PR. After looking deeper into the math and IEEE 754 standards, returning inf here is actually the correct and expected behavior. The true mathematical result of softplus(1.0, beta=1e-6) is approximately 693,147, which vastly exceeds the float16 maximum of 65,504. The standard says that the operations should return as if performed in infinite and unbounded precision and then rounded to the dtype. Narrowing conversions that exceed the target type's maximum should overflow to inf rather than silently clamping, so no changes are needed here.
